machine-learning/solutions/seismic_salt_detection/trainer/model.py

```diff
@@ -50,74 +48,67 @@ def forward_key_to_export(estimator):
-                    estimatorSpec.predictions)
-        return estimatorSpec
+                    estimator_spec.predictions)
+        return estimator_spec
     return tf.estimator.Estimator(model_fn=model_fn2, config=config)
 
-def create_regressor(config, parameters):
-    """Creates a DNN regressor.
 
-    Defines features and builds an 'Estimator' with them.
+def _estimator_metrics(labels, predictions):
+    """Creates metrics for Estimator.
+
+    Metrics defined here can be used to evaluate the model (on evaluation
+    data) and also can be used to maximize or minimize their values during
+    hyper-parameter tunning.
 
     Args:
-        config: `RunConfig` object to configure the runtime of the `Estimator`.
-        parameters: Parameters passed to the job.
+        labels: Evaluation true labels.
+        predictions: Evaluation model predictions.
 
     Returns:
-        A configured and ready to use `tf.estimator.DNNRegressor`
+        A dictionary with the evaluation metrics
     """
+    pred_logistic = predictions['logistic']
+    pred_class = predictions['class_ids']
+    return {
+        'accuracy': tf.metrics.accuracy(labels, pred_class),
+        'auc': tf.metrics.auc(labels, pred_logistic),
+        'auc_pr': tf.metrics.auc(labels, pred_logistic, curve='PR'),
+        'precision': tf.metrics.precision(labels, pred_class),
+        'recall': tf.metrics.recall(labels, pred_class)}
 
-    # Mean and Standard Deviation Constants for normalization.
-    with file_io.FileIO(parameters.mean_path, mode='r') as f:
-        mean = pickle.load(f)
-    with file_io.FileIO(parameters.std_path, mode='r') as f:
-        std = pickle.load(f)
-
-    # Columns to be used as features.
-    hour = tf.feature_column.categorical_column_with_identity(
-        'hour', num_buckets=24)
-    hour = tf.feature_column.embedding_column(
-        hour, dimension=parameters.hour_embedding)
 
-    day = tf.feature_column.categorical_column_with_identity(
-        'day', num_buckets=7)
-    day = tf.feature_column.embedding_column(
-        day, dimension=parameters.day_embedding)
+def create_classifier(config, parameters):
+    """Creates a DNN classifier.
 
-    weather = [tf.feature_column.numeric_column(
-        'weather' + str(i),
-        normalizer_fn=(lambda x, i = i: (x - mean[i]) / std[i])
-    ) for i in range(constants.WEATHER_SIZE)]
+    Defines features and builds an 'Estimator' with them.
 
-    distribution = [tf.feature_column.numeric_column(
-        'distribution' + str(i)
-    ) for i in range(constants.DISTRIBUTION_SIZE)]
+    Args:
+        config: `RunConfig` object to configure the runtime of the `Estimator`.
+        parameters: Parameters passed to the job.
 
-    feature_cols = [hour, day] + weather + distribution
+    Returns:
+        A configured and ready to use `tf.estimator.DNNClassifier`
+    """
+    # Mean and Standard Deviation Constants for normalization.
+    depth_mean = np.float32(parameters.depth_mean)
```

Comment:
In my previous solution I had a data preparation script that would do all the complicated  data munging including generating means and stds for my 180 features.

https://github.com/GoogleCloudPlatform/professional-services/blob/master/examples/cloudml-energy-price-forecasting/data_preparation/data_prep.py

In this solution we don't need any complicated data munging that's why I didn't create such script (I think it'll be an overkill to have one just to generate these two numbers don't you think? that's why I decided just to let the user input them with the job )

Reply:
I agree with you that a script like that is overkill.  There should be a way to embed per-record preprocessing directives into a served model as a part of packaging it up after training.  ... I don't suppose anything like that exists?

Reply:
There is a TF extension for DataFlow but running a DataFlow job for this will be even a bigger overkill than a script.

I can create a script like in the last solution for these if you think that is the right way forward.

Reply:
I like your solution as written.  I'll follow up with you about it next time we talk to continue my own education and understand the toolset better.  Thanks for the chat, here.